### DISCUSSION TITLE
Todo list

### DISCUSSION THREAD

**Discussion by wills106** [wills106/homeassistant-solax-modbus#37] [Ideas]
- Dynamic times for Charger Times
- Revisit disabled by default sensors as a number of them return funny values.
- Workout why number & select entities don't show in the main Integration panel.
~~#40~~ Moved to Wiki
~~#42~~ Moved to Wiki
~~- Allow for setting [Unit address](https://github.com/wills106/homsassistant-solax-modbus/discussions/37#discussioncomment-2449759) up instead of hard coded to 1~~
- Move away from using Tick boxes for Setup, read the [Serial number](https://github.com/wills106/homsassistant-solax-modbus/discussions/37#discussioncomment-2458998) from the Inverter and configure based off that

**Comment by wills106** [comment]
Release [0.4.9](https://github.com/wills106/homsassistant-solax-modbus/releases/tag/0.4.9) should start to address some of the hidden sensors returning random values.

**Reply by stringkillersa** [reply]
@wills106 Modbus select unit adress other than "1" as an option would be welcome ;) But that would require that two installations can use the same IP-address and port. (The Waveshare ethernet=>RS485 modbus accepts several connections.

**Reply by infradom** [reply]
#wills106 : I am willing to make this modification (first in my fork), the only disadvantage will be that we will have yet another config parameter in the config dialog (the modbus address). Let me know if I should proceed .. I cannot test with 2 inverters however. @stringkillersa will have to do the testing with 2 inverters behind a waveshare.

**Reply by wills106** [reply]
Yes that's fine. Hopefully we can move the rest away from the config once it reads the Serial number.

Then we should be left with just a few options.
Port, Address, Uni, Serial Device, EPS etc

**Reply by stringkillersa** [reply]
@infradom I can absolutely do  the testing with unit adresses.

**Reply by wills106** [reply]
[0.4.13](https://github.com/wills106/homsassistant-solax-modbus/releases/tag/0.4.13) Contains the change to select different Unit Numbers.

**Reply by stringkillersa** [reply]
@wills106 was about to try the Unit adressing, however - this requires two setups with the same IP-adress (with different unit address), but currently this gives error "Device is already configured"

**Reply by wills106** [reply]
Is that even when you give them separate name's? Like SolaX-A & SolaX-B

**Reply by stringkillersa** [reply]
@wills106 Yes, they are namned Solax-1 and Solax-2 but refuses to use same ip.

**Reply by stringkillersa** [reply]
@wills106 Give me a hint what code to remove for testing

**Reply by wills106** [reply]
The short answer is I don't know?

I will have to try and read up on it, it could be a limitation within HA or we need to enable something in the `__init__.py` or `config_flow.py`?

I know your Inverter's aren't in a master / slave setup. But I could still add in the ability for it to read the slave Inverters. Just in case it recognises multiple Inverters are connected together internally in the Inverters?
Are they setup as Unit=1 and Unit=2?

**Reply by stringkillersa** [reply]
@wills106 Yes, they have unit number 1 and 2 and not master/slave

**Reply by wills106** [reply]
It could be that we are using some built in functions:

`from homeassistant.const import (**CONF_HOST**, CONF_NAME, CONF_PORT, CONF_SCAN_INTERVAL)`

**CONF_HOST** is the IP Address holder.

For other ones we are defining them ourselves:

`from .const import (DEFAULT_SERIAL_PORT, DEFAULT_MODBUS_ADDR)`

We could possibly swap from using CONF_HOST and defining it ourselves like CONF_HOST_ADDR I don't know if it will let us then have the same address or not?

**Reply by stringkillersa** [reply]
@wills106 worth trying

**Reply by infradom** [reply]
@stringkillersa @wills106 I have created a pull request with a modification that should solve this problem:
In this change, the unique_id is no longer the host name but a string that also contains the modbus_addr and tcp_port.
This new style unique_id is only used if non-default modbus_addr or tcp_port is used. The latter is done to remain compatible with existing lovelace dashboards

**Reply by infradom** [reply]
This is now implemented in the prerelease version. Can you test with 2 inverters? Some remarks from the gen4 installation wiki page apply if you use waveshare.

**Reply by stringkillersa** [reply]
@infradom Sorry, it still answers "Device is already configured" when I specify the same IP address for the second setup, despite different modbus unit and integration prefix.

**Reply by infradom** [reply]
Did you use the latest prerelease?
In the new version, if the modbus address is non-standard 1, the device id is op address extended with the modbus address.
Maybe a previous try created this new id. Using yet another modbus address could resolve this.

I will reverify the code. I could declare my single inverter twice by also adding the name to the id in the source code; I removed the latter since your modbus addresses are already different

**Reply by stringkillersa** [reply]
@infradom Upgraded today, grep gave me manifest.json:  "version": "0.4.13"
Will test again tomowwow with other Unit adress

**Reply by wills106** [reply]
Are you installing through HACS?

If so you need to turn on the beta version option.

0.4.13 / 0.4.13a are the latest "Production" versions, but for beta / pre-release we are on 0.5.0b5. 0.5.0b1 added the fix for multiple Inverter on same IP Adress.

**Reply by stringkillersa** [reply]
@infradom Yes, using HACS, Ah, though I did select beta - but i see I missed that. OK, have installed 0.5.0b5
Installation went ok,  (if I select other than unit 1 in the first installation, it fails with the second).
However, there is something strange going on with the fetching of data. 
Using a modbus TCP/IP-tool i can read from both inverters (selecting unit modbus numbers) and confirm through the serial number that 
I am speaking with one or the other.  But as soon as I start up a polling from HA, both inverters freezes the modbus.
Even if I deactivate the Solax integration and reset the Waveshare device - the modbus is frozen. I have to physically
remove cables from both inverters and plug them in again to get communication working again.
Having TWO integrations with different polltimes different Units (1 and 2) configured - but only ONE inverter with unit number 2
physically connected - gave strangely enough values for entities belonging to inverter 1 as well 2.  So, either the inverter2 answered
both request (unit 1 and 2), or the waveshare converter messed up the unit number.
I have played around with the waveshare multi host values (not much to play with though), but  nothing that changed any behavior.
Is it possible that the Solax-integration still have a unit "1" somewhere hard coded?

**Reply by infradom** [reply]
Thanks for the feedback, strange that you get answers from a non-connected inverter.
You absolutely need to enter a different name for  the second inverter (I saw a similar mixup when I used the same name and needed to remover both to resolve it)
 I could only test by declaring my inverter twice (with a trick).  
When I used 2 times the 1 address, I saw the same data twice on both instances, which is expected.
When I used modbus address 2 on one of the integrations, this second one's attributes remained unknown, which is expected

I will reverify the code, but I am still busy with another development, and in the coming days, I may not have much time, so I cant promise a solution quickly.

**Reply by stringkillersa** [reply]
@infradom No worries, I have two waveshare in my setup, but for the future it would be nice to just need one. 
But it is not a big issue. More for coming users...

**Reply by wills106** [reply]
@stringkillersa are you able to try out "0.5.0b7 Pre-release" does the new sensor return anything:

- group_read_test

Just wondering if the Inverter knows if an additional Inverter is connected or not?

**Reply by stringkillersa** [reply]
@wills106 Just tested. Both inverters return 0:0W:0W:0W:0A

**Reply by wills106** [reply]
Ok, so it must only return values when in a true Master / Slave setup.

Was worth a try though.

**Comment by wills106** [comment]
Looking at moving away from the various Tick Boxes used during initial Setup.
Might still need one for additional EPS configuration?

But was thinking of something like the following:

```
if seriesnumber[0:4] == 'L50E':
    model = gen2x1
    max_charge = 100

elif seriesnumber[0:4] == 'U50E':
    model = gen2x1
    max_charge = 50

elif seriesnumber[0:4] == 'H34T':
    model = gen4x3
    max_charge = 25

elif seriesnumber[0:4] == 'HUE5':
    model = gen3x1
    max_charge = 25

else:
    model = gen3x1
    max_charge = 20
```

This takes the first 4 digits of the serial number (seriesnumber[0:4]) and compares it against the known list of inverters from #26 It sets the model to Gen2, Gen4 etc and also the maximum charge / discharge allowed for the battery.

Instead of having three lots of the following:

```
["Battery Charge Max Current",
    "battery_charge_max_current",
    0x24,
    "f",
    {
    "min": 0,
    "max": 20, #or 25, 50, 100 etc
    "step": 0.1,
     "unit": ELECTRIC_CURRENT_AMPERE,
    }
],
```

We can replace it with using a variable:

```
["Battery Charge Max Current",
    "battery_charge_max_current",
    0x24,
    "f",
    {
    "min": 0,
    "max": maxcharge,
    "step": 0.1,
     "unit": ELECTRIC_CURRENT_AMPERE,
    }
],
```

This will reduce the code repetition, also when new / alternative Inverter models get discovered we only have to update the main IF statement.

I need to workout where to put this main IF statement so it works across `__init__.py` and `const.py` etc
When I first placed it in `const.py` it didn't seem to to anything.
I then placed it after the "seriesnumber" register in `__init__.py` and it correctly detected my Inverter as a H1E5 but I was unable to pass back "model = gen3x1" & "max_charge = 20" back to `const.py`

Needs further investigation...

**Reply by infradom** [reply]
I believe we can solve this with the following suggestion (needs to be refined)
Const.py contains a list of all entities in one big group, something like

GEN2 = 0x000é
GEN3 = 0x0004
GEN4 = 0x0008
X1   = 0x0100
X3   = 0x0200
Hybrid  = 0x0400
AC = 0x0800
EPS  = 0x1000
ALL = 0xFFFF

ALL_ENTITIES = { # imaginary example

  "grid_export_total": MyEntityDescription(
     name="Grid Export Total",
     key="grid_export_total",
     native_unit_of_measurement=ENERGY_KILO_WATT_HOUR,
     icon="mdi:solar-power",
     device_class=DEVICE_CLASS_ENERGY,
     state_class=STATE_CLASS_TOTAL_INCREASING,
     entity_registry_enabled_default=False,
     mask = GEN2 | GEN3 | GEN4 | X1  | X3
    ),
  "backup_nightcharge_upper_soc"] = MyEntityDescription(
     name="Backup Night Charge Upper SOC",
     key="backup_nightcharge_upper_soc",
     native_unit_of_measurement=PERCENTAGE,
     device_class=DEVICE_CLASS_BATTERY,
     mask = GEN4 | X1 | X4
    ),
 "eps_min_esc_soc": MyEntityDescription(
     name="EPS Min Esc SOC",
     key="eps_min_esc_soc",
     native_unit_of_measurement=PERCENTAGE,
     mask = GEN2 | GEN3 | GEN4 | EPS
    ),

This const.py declaration does not yet create the entities yet
The __init__.py code could then traverse this list and  based on the serial number creatre the real entities, taking into account the bitmask.
A hybrid inverter like the Gen4 X3 EPS would get the characteristic bit code:

       Inverter_type = X3 | HYBRID | GEN4 | EPS  #(in hex 0x14208)

Determining whether an entity needs to be created for this inverter can be done with a simple bitwise and of inverter_type with the entity's mask attribute. (simplified explanation)


I may have time tomorrow to try to make a proof of concept .... (or proof of bad idea)

**Reply by infradom** [reply]
Hello Will,

I succeeded in implementing this approach for the entities and the sensors. I did not yet try to implement this for select, number and button. 
This is not yet a pull request, but I would like you to have a look at this approach. First tests look fine on my X3 Gen4
Deriving the invertertype bitmask from the serial number should be easier now imho
See const.py and sensor.py changes in my test fork https://github.com/infradom/homsassistant-solax-modbus

**Reply by wills106** [reply]
I am starting to go through the `__init__.py` to mark what is Gen2 only and what's Gen3 etc
Should make it easier to sort out which sensors belong to which Gen etc

Also going through the `const.py` at the same time to see if we are declaring sensors in `__init__.py` but not exposing them, or maybe missing a Inverter type out.

![image](https://user-images.githubusercontent.com/18155231/162617062-41e2ea66-dca1-416b-8697-a4f98719fd06.png)
Just so you know, that comment refers to the sensor below and not above.

So in this case "meter_function" and not "meter_2_id" above the comment. Hope that make's sense?

**Reply by wills106** [reply]
Would the following work @infradom in `const.py`

```
    SolaXModbusSensorEntityDescription(
        name="Total Solar Energy",
        key="solar_energy_total",
        native_unit_of_measurement=ENERGY_KILO_WATT_HOUR,
        icon="mdi:solar-power",
        device_class=DEVICE_CLASS_ENERGY,
        state_class=STATE_CLASS_TOTAL_INCREASING,
        entity_registry_enabled_default=False,
        allowedtypes= GEN3 | GEN4,
    ),
    SolaXModbusSensorEntityDescription(
        name="Total Solar Energy",
        key="solar_energy_total_gen2",
        native_unit_of_measurement=ENERGY_KILO_WATT_HOUR,
        icon="mdi:solar-power",
        device_class=DEVICE_CLASS_ENERGY,
        state_class=STATE_CLASS_TOTAL_INCREASING,
        entity_registry_enabled_default=False,
        allowedtypes= GEN2,
    ),
```

So the first "Total Solar Energy" would work for X1 & X3 versions of Gen3 & Gen4
The second "Total Solar Energy" would work for the Gen2

**Reply by wills106** [reply]
@infradom my change seems to be working ok without breaking Gen3 functionality. (hopefully the X3 stuff still functions as it should) @matthewjporter is just testing out on the Gen2.
If it all works as it should do, we now have full Gen2 support for the Energy Dashboard.

Probably won't know till the end of tomorrow if it's functioning as intended.

**Reply by infradom** [reply]
Ok, I just pulled your version (main branch) and it seems to be ok on the GEN4 inverter.
I agree that  the entities you marked with a X3 related comment should also have the X3 flag

**Reply by infradom** [reply]
Yes, the suggested 2 declarations would work as you describe. Not specifying the X parameters or the EPS parameter means they dont matter (empty groups are treated as Pass/True in the evaluation)

**Reply by wills106** [reply]
"Total Solar Energy" works on the later Gen2 SK-SU U50ED and not the earlier U50EC at the moment #55

**Reply by infradom** [reply]
@wills106 In my development fork, today, I further prepared the code for deriveing invertertype bitmask from serial number.
I do an initial read of the serial number (holding_registers_0) to determine the type. 
I have not yet put in the final serial number prefixes, I believe you have a better view on this.
I also made _invertertype a property (rather cosmetic)
Should I already create a pull request, or will we you wait until 0.5 is released ?

**Reply by wills106** [reply]
So if I understand correctly the section in `__init_.py` at Line 121 will be something like:

```
    # derive invertertupe from seriiesnumber
    if   seriesnumber.startswith('U50EC'):  invertertype = GEN2 | X1 # Plain Gen2 50A missing "Total Solar Energy"
    elif seriesnumber.startswith('U50ED'):  invertertype = GEN2 | X1 | SU # Gen2 50A with "Total Solar Energy"
    elif seriesnumber.startswith('L50E'):  invertertype = GEN2 | X1 | TL # Gen2 100A assuming with "Total Solar Energy"
    elif seriesnumber.startswith('H1E5'):  invertertype = GEN3 | X1 # Earlier Gen3 X1
    elif seriesnumber.startswith('HUE5'):  invertertype = GEN3 | X1 # Later Gen3 X1 no actual difference from H1E5
    elif seriesnumber.startswith('H3UE'):  invertertype = GEN3 | X3
    elif seriesnumber.startswith('H34'):  invertertype = GEN4 | X3
    else: _LOGGER.error(f"unrecognized inverter type - serial number : {seriesnumber}")
```

Then in `const.py` for the different Gen2 options would the following work?

```
NUMBER_TYPES = [
    SolaxModbusNumberEntityDescription( name = "Battery Charge Max Current", # multiple versions depending on GEN
        key = "battery_charge_max_current",
        register = 0x24,
        fmt = "f",
        min_value = 0,
        max_value = 50,
        step = 0.1,
        unit_of_measurement = ELECTRIC_CURRENT_AMPERE,
        allowedtypes = GEN2, # would this need a not TL? like: GEN2 ~ TL
    ),
    SolaxModbusNumberEntityDescription( name = "Battery Charge Max Current", # multiple versions depending on GEN
        key = "battery_charge_max_current",
        register = 0x24,
        fmt = "f",
        min_value = 0,
        max_value = 100,
        step = 0.1,
        unit_of_measurement = ELECTRIC_CURRENT_AMPERE,
        allowedtypes = GEN2 & TL, # needs to be Gen2 with TL
    ),

SENSOR_TYPES: list[SolaXModbusSensorEntityDescription] = [ 
    SolaXModbusSensorEntityDescription(
        name="Total Solar Energy",
        key="solar_energy_total_gen2",
        native_unit_of_measurement=ENERGY_KILO_WATT_HOUR,
        icon="mdi:solar-power",
        device_class=DEVICE_CLASS_ENERGY,
        state_class=STATE_CLASS_TOTAL_INCREASING,
        entity_registry_enabled_default=True,
        allowedtypes= GEN2 | SU | TL, # Only want this sensor on the Gen2 that have this register, early U50EC is missing it, would it need to be and or (SU | TL & GEN2)? I might be getting confused over these allowedtpyes
    ),
```

**Reply by wills106** [reply]
Just so you know I have edited my response above a couple of times. Incase you are reading an older email version?
But I am done now.

**Reply by infradom** [reply]
The SU and TL flags you declare are not known in the current framework. I my most recent version (15 minutes ago), I have added a max_charge property to the hub code.
Please have a look at this latest code. 
Alternatively, we could create the SU and TL flags if you wish ,

**Reply by infradom** [reply]
My previous answer was not complete. I will complete my previous answer by editing it. 
The & of bitmasks is dangerous as it distroys teh other groups, &~ is safer , but I need some time to think first ...

**Reply by wills106** [reply]
Yea what you have done makes more sense than having 4 or 5 blocks of "battery_charge_max_current" & "battery_discharge_max_current"

Hopefully we can find the missing "Total Solar Energy" for the earlier U50EC so we don't need to remove that sensor for the 1 Inverter type.

**Reply by infradom** [reply]
I dont like adding bitmask groups for a single inverter, for that reason, I would prefer to give the hub an additional property called "blacklist", that contains the keys of  the entities that need not  be created.

   if   seriesnumber.startswith('U50EC'):  
       invertertype = GEN2 | X1 
       blacklist[solar_enery_total] = True

**Reply by wills106** [reply]
As the max_value has already been set at 20, I don't need to declare it for the Gen3 X1 do I?
Also would the following OR work for the Gen3 X1 H1E5 or HUE5

```
    invertertype = 0
    max_value = 20 # safe default

    if seriesnumber.startswith('L50E'):
        invertertype = GEN2 | X1
        max_charge   = 100
    elif seriesnumber.startswith('U50E'):
        invertertype = GEN2 | X1
        max_charge   = 50
    elif seriesnumber.startswith('H1E5') | ('HUE5'): # Detecting either Gen3 X1 as the same Inverter also skipping max_charge
        invertertype = GEN3 | X1
    elif seriesnumber.startswith('H34T'):  
        invertertype = GEN4 | X3
        max_charge = 25
    else: 
        _LOGGER.error(f"unrecognized inverter type - serial number : {seriesnumber}")
```

Sorry if all these questions seam really basic.

**Reply by infradom** [reply]
indeed, but it shoud be 
seiresnumber.startswith('H1E5') or seriesnumber.startwith ('HUE5')

**Reply by wills106** [reply]
Ahh ok

**Reply by wills106** [reply]
> Should I already create a pull request, or will we you wait until 0.5 is released ?

Right I think I understand enough how this is going to work.

I think it's safe to add it in as a PR, then I can add the extra inverter types etc.

So the only time a user will need to fill in a Tick box is if their Inverter is an unknown Serial number or they need to enable the EPS function.

**Reply by infradom** [reply]
on second thought, I believe it is better to put the blacklist/exception list on the sensor declaration instaed of in the hub delclaration.
So give me some time to restructure slightly with some examples ...

**Reply by wills106** [reply]
That's fine, no rush.

**Reply by infradom** [reply]
I have introduced 2 new attributes in const.py: (in my fork )
- blacklist: for sensor, number, select, button entities
- max_exceptions: for number entities

The other files have been adapted accordingly
Please have a look if you like the approach, if so, I can issue a PR

**Reply by wills106** [reply]
I have noticed a couple of small corrections I need to do between sensors. But I understand what's going on now and I can correct them following the PR.

**Reply by infradom** [reply]
I will fist convert the blacklist and max_exception dictionarys to lists and then send the PR

**Reply by wills106** [reply]
Quick question, whats the difference between `allowedtypes = X1 | X3 | GEN2 | GEN3` and `allowedtypes = GEN2 | GEN3`
I was expecting them to be the same?

**Comment by infradom** [comment]
Yes, it is the same, same result, different approach
I prefer the shortest form now.

**Reply by wills106** [reply]
Ok, I just wanted to double check.
I knew `allowedtypes = X3 | GEN2 | GEN3` and `allowedtypes = X1 | GEN2 | GEN3` would result in one being for the X3 and the other for X1
But seeing `allowedtypes = X1 | X3 | GEN2 | GEN3` and `allowedtypes = GEN2 | GEN3` threw me off.

**Reply by infradom** [reply]
Sorry for the confusing approaches to reach the same .
ALLDEFAULTS could be simply 0 also

**Reply by infradom** [reply]
To be very clear: 
allowedtypes = X1 | X3 | GEN2 | GEN3
means: this entity is valid for (X1 or X3)  but only for (GEN2 or GEN3)

But as empty groups evaluate to True,  GEN2  | GEN3 is the same
X1 X3 belong to a group
GEN2 GEN3 GEN4 belong to another group
HYBRID and AC belong to yet another group
EPS is a group on its own, just a flag

**Reply by wills106** [reply]
I have this running on mine correctly now, after remembering to add the Inverter serial numbers into `__init__.py`... Opps

I have added in the X1-AC as a GEN3 | X1 for now, so we can test it. Will need to revisit it though so I can remove the PV sensors which don't exist.

**Reply by infradom** [reply]
@wills106 we may need to explicitly add HYBRID | AC to all sensors, except for the AC only sensors

**Reply by wills106** [reply]
There won't be any AC only sensor as such. When I created the packaged version I basically removed all the PV parts.

Or did you mean something like:

```
    SolaXModbusSensorEntityDescription(
        name="PV Current 1",
        key="pv_current_1",
        native_unit_of_measurement=ELECTRIC_CURRENT_AMPERE,
        device_class=DEVICE_CLASS_CURRENT,
        allowedtypes=HYBRID, # Only on all Hybrid Inverter
    ),
```

But something else like

```
    SolaXModbusSensorEntityDescription(
        name="E Charge Total",
        key="e_charge_total",
        native_unit_of_measurement=ENERGY_KILO_WATT_HOUR,
        icon="mdi:solar-power",
        device_class=DEVICE_CLASS_ENERGY,
        state_class=STATE_CLASS_TOTAL_INCREASING,
        entity_registry_enabled_default=False,
        allowedtypes= GEN3 | GEN4,
    ),
```

Would then turn into:

```
    SolaXModbusSensorEntityDescription(
        name="E Charge Total",
        key="e_charge_total",
        native_unit_of_measurement=ENERGY_KILO_WATT_HOUR,
        icon="mdi:solar-power",
        device_class=DEVICE_CLASS_ENERGY,
        state_class=STATE_CLASS_TOTAL_INCREASING,
        entity_registry_enabled_default=False,
        allowedtypes= HYBRID | AC | GEN3 | GEN4, # As this sensor is on Hybrid and AC Gen3 & Gen4
    ),
```

Hope that makes sense.

**Reply by infradom** [reply]
Yes, second option 

But Maybe it would ve simpler to call the flag NO_PV instead of 2 flags

**Reply by infradom** [reply]
I noticed you forgot to remove the old block of code:
```

   # remove following block by scanning on serial number in the block above
    if read_gen2x1: invertertype = GEN2 | X1
    if read_gen3x1: invertertype = GEN3 | X1
    if read_gen4x1: invertertype = GEN4 | X1
    if read_gen3x3: invertertype = GEN3 | X3
    if read_gen4x3: invertertype = GEN4 | X3
    invertertype = invertertype | HYBRID  # adapt for AC model
    # end of block
```
As long as this is there, we do not use the new model:

**Reply by infradom** [reply]
We also need to add the | HYBRID flag to all inverters, except the AC one in this code:
```
    # derive invertertupe from seriiesnumber
    if   seriesnumber.startswith('L50E'):  invertertype = GEN2 | X1 # Gen2 X1 SK-TL
    elif seriesnumber.startswith('U50E'):  invertertype = GEN2 | X1 # Gen2 X1 SK-SU
    elif seriesnumber.startswith('H1E5'):  invertertype = GEN3 | X1 # Gen3 X1 Early
    elif seriesnumber.startswith('HUE5'):  invertertype = GEN3 | X1 # Gen3 X1 Late
    elif seriesnumber.startswith('XAC36'):  invertertype = GEN3 | X1 # Needs adapting to AC Only in future, as no PV Sensors
    elif seriesnumber.startswith('H3UE'):  invertertype = GEN3 | X3 # Gen3 X3
    elif seriesnumber.startswith('F3E'):  invertertype = GEN3 | X3 # RetroFit
    elif seriesnumber.startswith('H34T'):  invertertype = GEN4 | X3 # Gen4 X3
    elif seriesnumber.startswith('H450'):  invertertype = GEN4 | X1 # Gen4 X1
    # add cases here
    #
    #
    else: 
        _LOGGER.error(f"unrecognized inverter type - serial number : {seriesnumber}")

```

**Reply by wills106** [reply]
I thought it was optional as when I tried it first without adding https://github.com/wills106/homsassistant-solax-modbus/blob/7e9c498435f1f8010f8395274e8a6692e9ef88af/custom_components/solax_modbus/__init__.py#L125

I got the _LOGGER.error(f"unrecognized inverter type - serial number : {seriesnumber}")

But then when I added in all the extra Inverter serial number I got my normal 111 sensors like I would do if I ticked Gen3 X1
But I didn't select any Tick Boxes.

**Reply by wills106** [reply]
I just removed the integration and readded it ticking Gen2 X1 and I am back down to 82 sensors.

So doesn't it function as a manual override? It also adopted the Gen2 scaling like it should.

**Reply by infradom** [reply]
yes, but the old code was still there ... masking the error
I updated my fork

**Reply by wills106** [reply]
O I thought it was intentional until we know it's working correctly before pushing out the production 0.5.0

**Reply by infradom** [reply]
indeed, but the new code is normally overriden by the old block. 
Maybe I went too fast ...and decided to test without the old block
Sorry

**Reply by infradom** [reply]
anyway, in the new block we need to add the HYBRID flag on most inverters. Here the rules for a sensor declaration do NOT apply (empty group rule)

**Reply by wills106** [reply]
I'll merge your PR but I think it's safe to keep 0.5.0b6 at the moment.

That code block just works as an override.

I changed to Gen4 X1 just, lost some Gen3 selects and gained sensors.

As long as people don't use the Tick boxes it just works based on the Serial number.

**Reply by infradom** [reply]
EDITED: ignore this comment
I am less sure about that, maybe I overlook something, but if no box is ticked, it will almost behave like a GEN3 in the code with the old block

**Reply by infradom** [reply]
OK, you are right, if no other box is ticked, the old code will do nothing

**Reply by infradom** [reply]
sorry for the confusion, you are right,

**Reply by wills106** [reply]
It's ok :)

**Reply by infradom** [reply]
you could add the block of code again (except the last line)
As long as the tick boxes are there, it can function as a fallback in case we forgot some SN codes

**Reply by wills106** [reply]
I did add the following to the 0.5.0b6 release notes:

```
If the Integration fails to load with the following error in your log “unrecognized inverter type - serial number : {your_serial_number_here}”
Please see [#26](https://github.com/wills106/homsassistant-solax-modbus/discussions/26) providing the details asked for.
```

As I wasn't sure at the time if the Tick Boxes worked or not.
Need to ideally start removing them at some point though.
